google/cloud/bigtable/internal/async_retry_unary_rpc.h

```diff
@@ -170,6 +170,183 @@ class EmptyResponseAdaptor {
+  using Request = RequestType;
+  using Response = ResponseType;
+  //@}
+
+  /**
+   * Start the asynchronous retry loop.
+   *
+   * @param location typically the name of the function that created this
+   *     asynchronous retry loop.
+   * @param rpc_retry_policy controls the number of retries, and what errors are
+   *     considered retryable.
+   * @param rpc_backoff_policy determines the wait time between retries.
+   * @param idempotent_policy determines if a request is retryable.
+   * @param metadata_update_policy controls how to update the metadata fields in
+   *     the request.
+   * @param async_call the callable to start a new asynchronous operation.
+   * @param request the parameters of the request.
+   * @param cq the completion queue where the retry loop is executed.
+   * @return a future that becomes satisfied when (a) one of the retry attempts
+   *     is successful, or (b) one of the retry attempts fails with a
+   *     retryable error, or (c) one of the retry attempts fails with a
+   *     retryable error, but the request is non-idempotent, or (d) the retry
+   *     policy is expired.
+   */
+  static future<StatusOr<Response>> Start(
+      char const* location, std::unique_ptr<RPCRetryPolicy> rpc_retry_policy,
+      std::unique_ptr<RPCBackoffPolicy> rpc_backoff_policy,
+      IdempotencyPolicy idempotent_policy,
+      MetadataUpdatePolicy metadata_update_policy, AsyncCallType async_call,
+      Request request, CompletionQueue cq) {
+    std::shared_ptr<RetryAsyncUnaryRpcFuture> self(new RetryAsyncUnaryRpcFuture(
+        location, std::move(rpc_retry_policy), std::move(rpc_backoff_policy),
+        std::move(idempotent_policy), std::move(metadata_update_policy),
+        async_call, std::move(request)));
+    auto future = self->final_result_.get_future();
+    self->StartIteration(self, cq);
+    return future;
+  }
+
+ private:
+  // The constructor is private because we always want to wrap the object in
+  // a shared pointer. The lifetime is controlled by any pending operations in
+  // the CompletionQueue.
+  RetryAsyncUnaryRpcFuture(char const* location,
+                           std::unique_ptr<RPCRetryPolicy> rpc_retry_policy,
+                           std::unique_ptr<RPCBackoffPolicy> rpc_backoff_policy,
+                           IdempotencyPolicy idempotent_policy,
+                           MetadataUpdatePolicy metadata_update_policy,
+                           AsyncCallType async_call, Request request)
+      : location_(location),
+        rpc_retry_policy_(std::move(rpc_retry_policy)),
+        rpc_backoff_policy_(std::move(rpc_backoff_policy)),
+        idempotent_policy_(std::move(idempotent_policy)),
+        metadata_update_policy_(std::move(metadata_update_policy)),
+        async_call_(std::move(async_call)),
+        request_(std::move(request)) {}
+
+  /// The callback for a completed request, successful or not.
+  void OnCompletion(std::shared_ptr<RetryAsyncUnaryRpcFuture> self,
```

Comment:
`StartIteration/OnCompletion()` look a little funny, where `self.get() == this` and we sometimes refer to, say, `rpc_retry_policy_` and sometimes `self->rpc_retry_policy_`.  Perhaps they should be `static` and use `self` exclusively?

Reply:
Good idea, thanks! Done (meaning they are `static` and use `self` exclusively).